src/integer_mod_q/polynomial_ring_zq/set.rs

```diff
@@ -0,0 +1,171 @@
+        let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        assert!(poly_ring.set_coeff(i8::MAX, 2).is_ok());
+        assert!(poly_ring.set_coeff(i16::MAX, 2).is_ok());
+    }
+
+    /// Ensure that the max of [`u8`] and [`u16`] works as a coefficient.
+    #[test]
+    fn set_unsigned_coeff() {
+        let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        assert!(poly_ring.set_coeff(u8::MAX, 2).is_ok());
+        assert!(poly_ring.set_coeff(u16::MAX, 2).is_ok());
+    }
+
+    /// Ensure that a general case is working.
+    #[test]
+    fn set_coeff_working() {
+        let modulus = ModulusPolynomialRingZq::from_str("5  1 0 0 1 5 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("3  0 2 1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+        let value = 10000;
+
+        poly_ring.set_coeff(0, value).unwrap();
+        poly_ring.set_coeff(2, value).unwrap();
+
+        let cmp_poly = PolyOverZ::from_str("3  10000 2 10000").unwrap();
+        let cmp_poly_ring = PolynomialRingZq::from((&cmp_poly, &modulus));
+        assert_eq!(cmp_poly_ring, poly_ring);
+    }
+
+    /// Ensure that the set polynomial is reduced correctly.
+    #[test]
+    fn set_reduce() {
+        let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("1  1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        poly_ring.set_coeff(3, 1).unwrap();
+
+        let cmp_poly = PolyOverZ::from_str("0").unwrap();
+        let cmp_poly_ring = PolynomialRingZq::from((&cmp_poly, &modulus));
+        assert_eq!(cmp_poly_ring, poly_ring);
+    }
+
+    /// Ensure that the correct coefficient is set and all others are set to `0`.
+    #[test]
+    fn set_coeff_rest_zero() {
+        let modulus = ModulusPolynomialRingZq::from_str("7  1 8 0 0 1 0 12 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("2  0 2").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        poly_ring.set_coeff(5, 10).unwrap();
+
+        let cmp_poly = PolyOverZ::from_str("6  0 2 0 0 0 10").unwrap();
+        let cmp_poly_ring = PolynomialRingZq::from((&cmp_poly, &modulus));
+        assert_eq!(cmp_poly_ring, poly_ring);
+    }
+
+    /// Ensure that the negative indices return an error.
+    #[test]
+    fn set_min_negative_coeff() {
+        let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        assert!(poly_ring.set_coeff(i64::MIN, 2).is_err());
```

Comment:
Also add a negative index check in the in the order of the degree.
```suggestion
    fn set_negative_index() {
        let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
        let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));

        assert!(poly_ring.set_coeff(-1, 2).is_err());
        assert!(poly_ring.set_coeff(i64::MIN, 2).is_err());
```